pennylane/devices/experimental/custom_device_3_numpydev/python_mps.py

```diff
@@ -0,0 +1,304 @@
+    #     C = np.tensordot(op_j, C, axes=(1, 1))  # j [j*], vR* [j] vR
+    #     C = np.tensordot(B.conj(), C, axes=([0, 1, 2], [1, 0, 2])) # [vL*] [j*] [vR*], [j] [vR*] [vR]
+    #     return C
+
+def split_truncate_theta(theta, chi_max, eps):
+    """Split and truncate a two-site wave function in mixed canonical form.
+
+    Split a two-site wave function as follows::
+          vL --(theta)-- vR     =>    vL --(A)--diag(S)--(B)-- vR
+                |   |                       |             |
+                i   j                       i             j
+
+    Afterwards, truncate in the new leg (labeled ``vC``).
+
+    Parameters
+    ----------
+    theta : np.Array[ndim=4]
+        Two-site wave function in mixed canonical form, with legs ``vL, i, j, vR``.
+    chi_max : int
+        Maximum number of singular values to keep
+    eps : float
+        Discard any singular values smaller than that.
+
+    Returns
+    -------
+    A : np.Array[ndim=3]
+        Left-canonical matrix on site i, with legs ``vL, i, vC``
+    S : np.Array[ndim=1]
+        Singular/Schmidt values.
+    B : np.Array[ndim=3]
+        Right-canonical matrix on site j, with legs ``vC, j, vR``
+    """
+    chivL, dL, dR, chivR = theta.shape
+    theta = np.reshape(theta, [chivL * dL, dR * chivR])
+    X, Y, Z = svd(theta, full_matrices=False) #TODO switch to qml.math.svd
+    # truncate
+    chivC = min(chi_max, np.sum(Y > eps))
+    assert chivC >= 1
+    piv = np.argsort(Y)[::-1][:chivC]  # keep the largest `chivC` singular values
+    X, Y, Z = X[:, piv], Y[piv], Z[piv, :]
+    # renormalize
+    S = Y / np.linalg.norm(Y)  # == Y/sqrt(sum(Y**2))
+    # split legs of X and Z
+    A = np.reshape(X, [chivL, dL, chivC])
+    B = np.reshape(Z, [chivC, dR, chivR])
+    return A, S, B
+
+def update_bond(psi, i, U_bond, chi_max, eps):
+    """Apply `U_bond` acting on i,j=(i+1) to `psi`."""
+    j = (i + 1) % psi.L
+    # construct theta matrix
+    theta = psi.get_theta2(i)  # vL i j vR
+    # apply U
+    Utheta = np.tensordot(U_bond, theta, axes=([2, 3], [1, 2]))  # i j [i*] [j*], vL [i] [j] vR
+    Utheta = np.transpose(Utheta, [2, 0, 1, 3])  # vL i j vR
+    # split and truncate
+    Ai, Sj, Bj = split_truncate_theta(Utheta, chi_max, eps)
+    # put back into MPS
+    Gi = np.tensordot(np.diag(psi.Ss[i]**(-1)), Ai, axes=(1, 0))  # vL [vL*], [vL] i vC
+    psi.Bs[i] = np.tensordot(Gi, np.diag(Sj), axes=(2, 0))  # vL i [vC], [vC] vC
+    psi.Ss[j] = Sj  # vC
+    psi.Bs[j] = Bj  # vC j vR
+    return psi
+
+def update_site(psi, i, U_site, chi_max, eps):
+    """Apply `U_site` acting on site i to `psi`"""
+    # Conctract Theta and U
+    # restore B-form by left-multiplying inverse singular values
+    theta = psi.get_theta1(i)
+    Utheta = np.tensordot(U_site, theta, axes=([1], [1]))  # i [i*], vL [i] vR
+    Utheta = np.transpose(Utheta, [1, 0, 2])  # i vL vR >> vL i vR
+
+    # no truncation (unsure atm)
```

Comment:
I'm a bit confused here, do I need to do any SVD for single site updates? In single-site DMRG yes, but here this is different. Maybe @chaeyeunpark knows about this?

Reply:
I will have a look in the new year. I also want to tag @ikurecic here.

Reply:
I know now it is wrong, curious to know how to do it for single site (atm I am just forcing a two-site update)